src/lib/clients/main-client.ts

```diff
@@ -0,0 +1,365 @@
+      throw new Error(`[${validAuthData.error}] ${validAuthData.error_description}`)
+    }
+
+
+    // TODO – parse the state key from localStorage
+    return validAuthData;
+  }
+
+  private buildQueryParams (options: LoginOptions): OAuthRequestParams {
+    const { redirectUri, org, provider, state } = options;
+
+    const query: OAuthRequestParams = {
+      client_id: encodeURIComponent(this.config.clientId),
+      response_type: 'token'
+    };
+
+    if (redirectUri) query.redirect_uri = encodeURIComponent(redirectUri);
+    if (org) query.org = encodeURIComponent(org);
+    if (provider) query.provider = encodeURIComponent(provider);
+    if (state) {
+      let stateString: string = state as any;
+      if (typeof state !== 'string') {
+        stateString = this.saveStateToStorage(state);
+      }
+      query.state = encodeURIComponent(stateString);
+    }
+
+    return query;
+  }
+
+  private readStateInStorage (id: string): POJO | undefined {
+    const item = localStorage.getItem(id);
+    return item ? JSON.parse(item) : undefined;
+  }
+
+  private deleteStateInStorage (id: string): void {
+    localStorage.removeItem(id);
+  }
+
+  private saveStateToStorage (state: POJO, id?: string): string {
+    const storageId = id || this._generateId();
+
+    debug('writing state data', { storageId, state });
+    localStorage.setItem(storageId, JSON.stringify(state));
+
+    return storageId;
+  }
+
+  /**
+   * Return a unique id for client-auth
+   * @param id optional id string
+   * @returns unique id specific to client-auth
+   */
+  private _generateId (id?: string): string {
+    return `gc-ca_${id || v4()}`;
+  }
+
+  /**
+   * Clears auth data from localStorage and redirects the user to the GenesysCloud logout page
+   *
+   * @param logoutRedirectUri optional, redirectUri to pass to the logout page
+   */
+  logout (logoutRedirectUri?: string): void {
+    this.clearAuthData();
+
+    const query: { [key: string]: string } = {
+      client_id: encodeURIComponent(this.config.clientId)
+    };
+
+    if (logoutRedirectUri) {
+      query.redirect_uri = encodeURIComponent(logoutRedirectUri);
+    }
+
+    const url = this.buildAuthUrl('logout', query);
+    window.location.replace(url);
+  }
+
+  /**
+ * Sets the access token on the authenticator instance and localStorage (if configured)
+ * @param token - The access token
+ */
+  setAccessToken (token: string): void {
+    this._writeAuthData({ accessToken: token });
+  }
+
+  async validateAuthData (
+    incomingData: AuthData,
+    dataFromStorage: AuthData
+  ): Promise<AuthData | undefined> {
+    const tests = [
+      { /* test incoming data first */
+        data: incomingData,
+        debugMsg: 'incoming auth data (via passed in param or from window hash) failed token test'
+      },
+      {
+        data: dataFromStorage,
+        debugMsg: 'auth data from localStorage failed token test'
+      }
+    ] as const;
+
+    for (let test of tests) {
```

Comment:
Codacy has a fix for the issue: ['test' is never reassigned. Use 'const' instead.](https://app.codacy.com/gh/MyPureCloud/genesys-cloud-client-auth/pullRequest?prid=9713744)
```suggestion
    for (const test of tests) {
```